src/Renci.SshNet/Common/SshDataStream.cs

```diff
@@ -264,5 +276,21 @@ private byte[] ReadBytes(int length)
 
             return data;
         }
+
+#if NETSTANDARD2_1 || NET6_0_OR_GREATER
+        /// <summary>
+        /// Reads data into the specified <paramref name="buffer" />.
+        /// </summary>
+        /// <param name="buffer">The buffer to read into.</param>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="buffer"/> is larger than the total of bytes available.</exception>
+        private void ReadBytes(Span<byte> buffer)
+        {
+            var bytesRead = Read(buffer);
+            if (bytesRead < buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(buffer), string.Format(CultureInfo.InvariantCulture, "The requested length ({0}) is greater than the actual number of bytes read ({1}).", buffer.Length, bytesRead));
```

Comment:
Wouldn't it be better to throw exceptions from separate methods? I don't remember exactly, but it was probably better for performance reasons. Throwing exceptions probably blocked the method inlining.

This doesn't block this PR, it's more of a question.

Reply:
The benefit of using "throw helpers" is that the code size of the method is reduced by shifting the "cold" throwing code into another method (which is not inlined).

See [this example](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdACJQQDmMA9iglAMYpUDCzAJgFMAgjAgAbAJ4ooKANxFSFagHFxzYBKgAvCO2YwFxcpSoBJAPJHMAZgrYyPMiDIBZQWGYAnSQGUEXoIQYGQA3kRkkRR2mAAsZABKQfwAQpIIgii4ABS+AA4QMAA8wOmCAHxkwACuAGa1gl4AlGERUe0AbhBeVWUoSRD8ZAC8icnZNfWNTUbt7VC1ZBN9A0NFVXUNXlQAMoIwDAgAFk1tc62E5+fHXswA7mQwgg/CXgzVYPsIFtXftQmFBiCACiAA8OII8voYNkxJ9mLUJptpnBlCQqAAxbyQBDZHjVcQIaqBMwwWrMcwwLpeegwBD4wnEwSogBEABUjoIyIEAI7VTIZIbifaHI5LUIkAC+LVkZAYgT0jTIx0Kys5ZAgHCJEkeH2ASoRvQyKG5yXFuGlVBZqMmW12IuONpWySaMzOc0l7s9lyi7tsFHiqzSxuwuQKxVKGUqtumFyuZGpRsyqxGY0GSKmzVmVwWS0jybN6xj2z2B2Opx953ClfjZA5tzuGa2TuNqzdNfa3vj7vaAG0aMxMgA5ZgIJJErwwAC6PailAAbAG60cG2HCiUytHkV5UbAEEn+i7Z5Fq7Wojd7o9nmRXu9PnSfn8AQcQeDIdDYcFBAimyi0ZjsT0PECQnQRSXJSlqVpekQKZVkOS5Xl+TYQQhQdMVsglaUyFleUggyHoVRgNUuU1bVxF1MB9R6Q18xNBUhkwi0mitG1t3tMsjhbAtBldbNzklbsa29SUgA). The generated code for `ReadBytes2` is much smaller than `ReadBytes1`.

In practice, I don't think we need to be thinking at this level of optimisation just yet